### PR TITLE
feat: offline adding

### PR DESCRIPTION
**Pull request by b5**
I have no idea what this does for our provide story if a user adds content offline then tries to take it online, but this makes our perf measurable.

I still think we should provide the root to the DHT by default. It's usually what users are trying to do, and while it's slower, it's the de-facto use case

### CODE REVIEW

File: iroh/src/run.rs [hunk 1]
```diff
@@ -175,13 +186,35 @@ async fn add(api: &impl Api, path: &Path, no_wrap: bool, recursive: bool) -> Res
...
     }
 
+    let mut steps = 3;
     // we require p2p for adding right now because we don't have a mechanism for
     // hydrating only the root CID to the p2p node for providing if a CID were
     // ingested offline. Offline adding should happen, but this is the current
     // path of least confusion
-    require_services(api, HashSet::from(["store", "p2p"])).await?;
+    let svc_status = require_services(api, HashSet::from(["store"])).await?;
+    match (provide, svc_status.p2p.status()) {
```
Comment (rklaehn) [inline iroh/src/run.rs:195]: Not for this PR, but we have to make sure that we refuse to speak to services with different build version

File: iroh/src/run.rs [hunk 2]
```diff
@@ -175,13 +186,35 @@ async fn add(api: &impl Api, path: &Path, no_wrap: bool, recursive: bool) -> Res
...
+        (true, ServiceStatus::Down(_status)) => {
+            return Err(anyhow!("Add provides content to the IPFS network by default, but the p2p service is not running.\n{}",
+            "hint: try using the --offline flag, or run 'iroh start p2p'".yellow()
+            ));
+        }
+        (true, ServiceStatus::Unknown)
+        | (true, ServiceStatus::NotServing)
+        | (true, ServiceStatus::ServiceUnknown) => {
+            return Err(anyhow!("Add provides content to the IPFS network by default, but the p2p service is not running.\n{}",
```
Comment (rklaehn) [inline iroh/src/run.rs:204]: There is anyhow::bail! which does this and maybe looks a bit nicer.

https://docs.rs/anyhow/latest/anyhow/macro.bail.html

File: iroh/src/services.rs [hunk 3]
```diff
@@ -252,8 +252,12 @@ where
     Ok(())
 }
 
-/// require a set of services is up
-pub async fn require_services(api: &impl Api, services: HashSet<&str>) -> Result<()> {
+/// require a set of services is up. returns the underlying status table of all
+/// services for additional scrutiny
+pub async fn require_services(
+    api: &impl Api,
+    services: HashSet<&str>,
```
Comment (rklaehn) [inline iroh/src/services.rs:259]: I would prefer BTreeSet in such cases because it is more deterministic. But in this case it does not really matter.

**Review by rklaehn — APPROVED** [review]
Looks good, except for the `return Err(anyhow!`.

File: iroh/src/run.rs [hunk 4]
```diff
@@ -161,7 +166,13 @@ impl Cli {
     }
 }
 
-async fn add(api: &impl Api, path: &Path, no_wrap: bool, recursive: bool) -> Result<()> {
+async fn add(
+    api: &impl Api,
+    path: &Path,
+    no_wrap: bool,
+    recursive: bool,
+    provide: bool,
```
Comment (dignifiedquire) [inline iroh/src/run.rs:174]: the next person touching this needs to create a config struct

File: iroh/src/services.rs (hunk 3 above)
Comment (b5) [reply iroh/src/services.rs:259]: yeah totally. I'll come through & refactor all uses of `HashSet` to `BTreeSet` in a follow up

**Pull request merged**